server/src/main/java/org/opensearch/indices/replication/SegmentReplicationSourceHandler.java

```diff
@@ -159,6 +169,7 @@ public synchronized void sendFiles(GetSegmentFilesRequest request, ActionListene
 
             sendFileStep.whenComplete(r -> {
                 try {
+                    this.updateVisibleCheckpointForShard(allocationId, copyState.getCheckpoint());
```

Comment:
I think we need to stop the timer & log outside of the finally block on line 176 or we will potentially stop the timer twice.  This won't break anything in a real case, but will break tests or when assertions are enabled & we concurrently close because the timer will be stopped twice.  I pulled this into https://github.com/opensearch-project/OpenSearch/issues/8292 and can see this happen.

```
jul 10, 2023 2:12:26 AM com.carrotsearch.randomizedtesting.RandomizedRunner$QueueUncaughtExceptionsHandler uncaughtException
WARNING: Uncaught exception in thread: Thread[#18598,opensearch[node_s2][generic][T#8],5,TGRP-SegmentReplicationSuiteIT]
java.lang.AssertionError: already stopped
	at __randomizedtesting.SeedInfo.seed([83C3411EF1371E24]:0)
	at org.opensearch.indices.replication.common.ReplicationTimer.stop(ReplicationTimer.java:82)
	at org.opensearch.indices.replication.SegmentReplicationSourceHandler.lambda$sendFiles$4(SegmentReplicationSourceHandler.java:141)
```
With `testDeleteIndexWhileReplicating`

Reply:
Updated as part of this PR.